File: algorithm/src/main/java/baekjoon/p09742/Main.java
package baekjoon.p09742;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (sc.hasNext()) {
            System.out.println(solve(sc.next(), sc.nextInt()));
        }
    }

    private static String solve(String input, int n) {
        char[] arr = input.toCharArray();
        Arrays.sort(arr);

        return String.format("%s %d = %s", input, n, solve2(list(arr), n));
    }

    private static List<char[]> list(char[] arr) {
        List<char[]> ret = new ArrayList<>();
        char[] next = arr;

        while (next != null) {
            ret.add(next);
            next = next(next);
        }

        return ret;
    }

    private static char[] next(char[] arr) {
        char[] ret = arr.clone();

        int idx = idx(ret);
        if (idx == -1) {
            return null;
        }

        int min = min(ret, idx);

        swap(ret, idx, min);
        Arrays.sort(ret, idx + 1, ret.length);

        return ret;
    }

    private static int idx(char[] arr) {
        for (int i = arr.length - 1; i > 0; i--) {
            if (arr[i] > arr[i - 1]) {
                return i - 1;
            }
        }

        return -1;
    }

    private static int min(char[] arr, int min) {
        for (int i = arr.length - 1; i > 0; i--) {
            if (arr[i] > arr[min]) {
                return i;
            }
        }

        return -1;
    }

    private static void swap(char[] arr, int a, int b) {
        char temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    private static String solve2(List<char[]> list, int n) {
        if (list.size() < n) {
            return "No permutation";
        } else {
            return String.valueOf(list.get(n - 1));
        }
    }

}

File: algorithm/src/main/java/baekjoon/p09742/Main.kt
package baekjoon.p09742

fun main(args: Array<String>) {
    while (true) {
        val input = readLine() ?: break
        val inputs = input.split(" ")
        println(solve(inputs[0], inputs[1].toInt()))
    }
}

private fun solve(input: String, n: Int): String {
    val arr = input.toCharArray().sorted().toCharArray()
    return "$input $n = ${solve2(list(arr), n)}"
}

private fun list(arr: CharArray): List<CharArray> {
    val ret = ArrayList<CharArray>()
    var next = arr

    while (next.isNotEmpty()) {
        ret.add(next)
        next = next(next)
    }

    return ret
}

private fun next(arr: CharArray): CharArray {
    val ret = arr.clone()
    val idx = (ret.lastIndex - 1 downTo 0).find { ret[it] < ret[it + 1] }
            ?: return emptyArray<Char>().toCharArray()
    val min = (ret.lastIndex downTo idx + 1).find { ret[it] > ret[idx] }!!

    swap(ret, idx, min)
    ret.sort(idx + 1)

    return ret
}

private fun swap(arr: CharArray, a: Int, b: Int) {
    val temp = arr[a]
    arr[a] = arr[b]
    arr[b] = temp
}

private fun solve2(list: List<CharArray>, n: Int): String {
    return when (list.size >= n) {
        true -> list[n - 1].joinToString("")
        false -> "No permutation"
    }
}